File: Codes/3904-smallest-stable-index-ii.java
/*
 * @Author: LetMeFly
 * @Date: 2026-09-05 08:24:28
 * @LastEditors: LetMeFly.xyz
 * @LastEditTime: 2026-09-05 08:49:55
 */
class Solution {
    public int firstStableIndex(int[] nums, int k) {
        int n = nums.length;
        int[] mini = new int[n];
        mini[n-1] = nums[n-1];
        for (int i = n - 2; i >= 0; i--) {
            mini[i] = Math.min(nums[i], mini[i + 1]);
        }
        for (int i = 0, M = 0; i < n; i++) {
            M = Math.max(M, nums[i]);
            if (M - mini[i] <= k) {
                return i;
            }
        }
        return -1;
    }
}

File: Codes/3904-smallest-stable-index-ii.rs
/*
 * @Author: LetMeFly
 * @Date: 2026-09-05 08:24:28
 * @LastEditors: LetMeFly.xyz
 * @LastEditTime: 2026-09-05 08:55:35
 */
impl Solution {
    pub fn first_stable_index(nums: Vec<i32>, k: i32) -> i32 {
        let n = nums.len();
        let mut mini = vec![0; n];
        mini[n-1] = nums[n-1];
        for i in (0..n-1).rev() {
            mini[i] = nums[i].min(mini[i+1]);
        }
        let mut M = 0;
        for i in 0..n {
            M = M.max(nums[i]);
            if M - mini[i] <= k {
                return i as i32;
            }
        }
        -1
    }
}
